sqlx-core/src/net/socket/mod.rs

```diff
@@ -192,53 +196,94 @@ pub async fn connect_tcp<Ws: WithSocket>(
     // IPv6 addresses in URLs will be wrapped in brackets and the `url` crate doesn't trim those.
     let host = host.trim_matches(&['[', ']'][..]);
 
-    #[cfg(feature = "_rt-tokio")]
-    if crate::rt::rt_tokio::available() {
-        use tokio::net::TcpStream;
+    let addresses = if let Ok(addr) = host.parse::<Ipv4Addr>() {
+        let addr = SocketAddrV4::new(addr, port);
+        vec![SocketAddr::V4(addr)].into_iter()
+    } else if let Ok(addr) = host.parse::<Ipv6Addr>() {
+        let addr = SocketAddrV6::new(addr, port, 0, 0);
+        vec![SocketAddr::V6(addr)].into_iter()
+    } else {
+        let host = host.to_string();
+        spawn_blocking(move || {
+            let addr = (host.as_str(), port);
+            ToSocketAddrs::to_socket_addrs(&addr)
+        })
+        .await?
+    };
+
+    let mut last_err = None;
+
+    // Loop through all the Socket Addresses that the hostname resolves to
+    for socket_addr in addresses {
+        match connect_tcp_address(socket_addr).await {
+            Ok(stream) => return Ok(with_socket.with_socket(stream).await),
+            Err(e) => last_err = Some(e),
+        }
+    }
 
-        let stream = TcpStream::connect((host, port)).await?;
-        stream.set_nodelay(true)?;
+    // If we reach this point, it means we failed to connect to any of the addresses.
+    // Return the last error we encountered, or a custom error if the hostname didn't resolve to any address.
+    Err(match last_err {
+        Some(err) => err,
+        None => io::Error::new(
+            io::ErrorKind::AddrNotAvailable,
+            "Hostname did not resolve to any addresses",
+        )
+        .into(),
+    })
+}
 
-        return Ok(with_socket.with_socket(stream).await);
-    }
+async fn connect_tcp_address(socket_addr: SocketAddr) -> crate::Result<impl Socket> {
+    cfg_if! {
+        if #[cfg(feature = "_rt-tokio")] {
+            if crate::rt::rt_tokio::available() {
+                use tokio::net::TcpStream;
 
-    #[cfg(feature = "_rt-async-std")]
-    {
-        use async_io::Async;
-        use async_std::net::ToSocketAddrs;
-        use std::net::TcpStream;
-
-        let mut last_err = None;
-
-        // Loop through all the Socket Addresses that the hostname resolves to
-        for socket_addr in (host, port).to_socket_addrs().await? {
-            let stream = Async::<TcpStream>::connect(socket_addr)
-                .await
-                .and_then(|s| {
-                    s.get_ref().set_nodelay(true)?;
-                    Ok(s)
-                });
-            match stream {
-                Ok(stream) => return Ok(with_socket.with_socket(stream).await),
-                Err(e) => last_err = Some(e),
+                let stream = TcpStream::connect(socket_addr).await?;
+                stream.set_nodelay(true)?;
+
+                Ok(stream)
+            } else {
+                crate::rt::missing_rt(socket_addr)
```

Comment:
This is actually a breaking behavior change. The existing behavior is to fall back to `async_io` if a Tokio runtime isn't available, which could be important for some use-cases.

Because of Cargo's [feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification), if a library using SQLx turns on the `runtime-tokio` feature, it's forced on for users of that library as well, whether they want it or not. The only way to turn it off is to patch the library.

This change would make a situation like that a hard blocker instead of mildly annoying.